src/main/scala/com/fulcrumgenomics/vcf/JointVariantContextIterator.scala

```diff
@@ -0,0 +1,65 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.vcf
+
+import htsjdk.samtools.SAMSequenceDictionary
+import htsjdk.variant.variantcontext.{VariantContext, VariantContextComparator}
+
+/**
+  * Iterates over multiple variant context iterators such that we return a list of contexts for the union of sites
+  * across the iterators.  If samples is given, we subset each variant context to just that sample.
+  */
+class JointVariantContextIterator(iters: Seq[Iterator[VariantContext]],
+                                  dict: SAMSequenceDictionary,
+                                  comp: Option[VariantContextComparator] = None,
+                                  sampleNames: Seq[String] = Seq.empty
+                                 )
+extends Iterator[Seq[Option[VariantContext]]] {
+
+  if (iters.isEmpty) throw new IllegalArgumentException("No iterators given")
+
+  private val iterators = iters.map(_.buffered)
+  private val comparator = comp.getOrElse(new VariantContextComparator(dict))
+
+  if (sampleNames.nonEmpty && iters.length != sampleNames.length) {
+    throw new IllegalArgumentException("Samples should have the same length as the number of iterators.")
+  }
+
+  def hasNext(): Boolean = iterators.exists(_.nonEmpty)
+
+  def next(): Seq[Option[VariantContext]] = {
```

Comment:
This seems fine, but it would probably be more efficient (and maybe more complicated) to keep the iterators in a sorted data structure  (e.g. TreeSet) using the comparator on the head for each one, then when you consume an element, pop the iterator out and back into the set.  That way you don't need to do a full sort each time through, but rather just re-sort those that have changed.